### PR TITLE
refactor and improve swagger documentation normalizer

### PR DESCRIPTION
**Pull request by piotrbrzezina**
| Q             | A
| ------------- | ---
| Bug fix?      |no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/api-platform/api-platform/issues/185
| License       | MIT
| Doc PR        | 

The solution of the problem which you presented in https://github.com/api-platform/api-platform/issues/185 requires to overwrite the entire class, which is very time-consuming.
That's why I'd like to propose a different mechanism for extraction of documentation from api resources. My solution allows to add individual mechanisms generating documentation

What do you think about my solution? @dunglas  @sroze @theofidry

### CODE REVIEW

File: src/Bridge/Symfony/Bundle/Resources/config/swagger.xml [hunk 1]
```diff
@@ -6,34 +6,141 @@
...
-            <argument type="service" id="api_platform.metadata.property.metadata_factory" />
-            <argument type="service" id="api_platform.resource_class_resolver" />
-            <argument type="service" id="api_platform.operation_method_resolver" />
-            <argument type="service" id="api_platform.operation_path_resolver" />
-            <argument type="service" id="api_platform.router" />
-            <argument type="service" id="api_platform.filters" />
-            <argument type="service" id="api_platform.name_converter" on-invalid="ignore" />
+        <!-- Extractors -->
+        <service id="api_platform.swagger.extractor.swagger_extractor"
+                 alias="api_platform.core.swagger.extractor.route_documentation_extractor"/>
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/swagger.xml]: extra space to add before `/>` according to our CS.

File: src/Bridge/Symfony/Bundle/Resources/config/swagger.xml [hunk 2]
```diff
@@ -6,34 +6,141 @@
...
 
-            <tag name="serializer.normalizer" priority="16" />
+        <service class="ApiPlatform\Core\Swagger\Extractor\RouteExtractor"
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/swagger.xml]: I think that you can make this service private (`public="false"`) to improve performance.

The `id` should come first  and use a self closing XML tag.

Generally speaking (for all XML files), we prefer to put the whole XML tag on only one line.

File: src/Swagger/Extractor/CollectionGetExtractor.php [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+use ApiPlatform\Core\Swagger\Model\SwaggerDocumentation;
+use ApiPlatform\Core\Swagger\Model\SwaggerMetadata;
+use ApiPlatform\Core\Swagger\Util\SwaggerDefinitions;
+use ApiPlatform\Core\Swagger\Util\SwaggerFilterDefinitions;
+use Symfony\Component\HttpFoundation\Request;
+
+class CollectionGetExtractor implements SwaggerExtractorInterface
+{
+    /**
+     * @var SwaggerExtractorInterface
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: Useless PHP doc, to remove (same for all those properties).

File: src/Swagger/Extractor/CollectionGetExtractor.php [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                $parameters = array_merge($parameters, $filtersParameters);
+            }
+
+            $documentation->setParameters($parameters);
+        }
+    }
+
+    private function paramTypeResolver($paramName)
+    {
+        if ($paramName === 'id') {
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: We use Yoda style (`'id' === $paramName`).

File: src/Swagger/Extractor/CollectionGetExtractor.php [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            $filtersParameters = $this->swaggerFilterDefinitions->get($swaggerMetadata);
+            if (count($filtersParameters) > 0) {
+                $parameters = array_merge($parameters, $filtersParameters);
+            }
+
+            $documentation->setParameters($parameters);
+        }
+    }
+
+    private function paramTypeResolver($paramName)
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: Please use scalar type hints and return types everywhere.

File: src/Swagger/Extractor/CollectionGetExtractor.php [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        }
+    }
+
+    /**
+     * @param SwaggerMetadata $swaggerMetadata
+     * @param $documentation
+     */
+    protected function extractParameters(SwaggerMetadata $swaggerMetadata, SwaggerDocumentation $documentation)
+    {
+        if (empty($documentation->getParameters())) {
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: Prefer using a guard clause (invert the test return early to reduce the level of indentation and the complexity of the block).

File: src/Swagger/Extractor/CollectionGetExtractor.php [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+    public function extract(SwaggerMetadata $swaggerMetadata): SwaggerDocumentation
+    {
+        if (null !== $this->decoratedExtractor) {
+            $documentation = $this->decoratedExtractor->extract($swaggerMetadata);
+        } else {
+            $documentation = new SwaggerDocumentation();
+        }
+
+        if ($swaggerMetadata->isCollection() && in_array(Request::METHOD_GET, $swaggerMetadata->getMethods())) {
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: Please use the string `GET` directly to not introduce a dependency to Symfony's HttpFoundation for this component.

File: src/Swagger/Extractor/CollectionGetExtractor.php [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
+use ApiPlatform\Core\Swagger\Model\SwaggerDocumentation;
+use ApiPlatform\Core\Swagger\Model\SwaggerMetadata;
+use ApiPlatform\Core\Swagger\Util\SwaggerDefinitions;
+use ApiPlatform\Core\Swagger\Util\SwaggerFilterDefinitions;
+use Symfony\Component\HttpFoundation\Request;
+
+class CollectionGetExtractor implements SwaggerExtractorInterface
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: We make all our classes final (and so, we don't use protected methods).

File: src/Swagger/Extractor/CollectionGetExtractor.php [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+     * @param SwaggerMetadata $swaggerMetadata
+     * @param $documentation
+     * @param $resourceMetadata
+     */
+    protected function extractResponses(
+        SwaggerMetadata $swaggerMetadata,
+        SwaggerDocumentation $documentation,
+        ResourceMetadata $resourceMetadata)
+    {
+        if (empty($documentation->getResponses())) {
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: Prefer using a guard clause (invert the test return early to reduce the level of indentation and the complexity of the block).

Comment (dunglas) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: By the way, empty is useless, we prefer `if (!$documentation->getResponses())`

File: src/Swagger/Extractor/CollectionGetExtractor.php [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+    /**
+     * @param SwaggerMetadata $swaggerMetadata
+     * @param $documentation
+     */
+    protected function extractParameters(SwaggerMetadata $swaggerMetadata, SwaggerDocumentation $documentation)
+    {
+        if (empty($documentation->getParameters())) {
+            $parameters = [];
+            if (!empty($swaggerMetadata->getRouteVariables())) {
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: Useless empty.

File: src/Swagger/Extractor/CollectionGetExtractor.php [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+                    $parameters[] = [
+                        'name' => $param,
+                        'in' => 'path',
+                        'required' => true,
+                        'type' => $this->paramTypeResolver($param),
+                    ];
+                }
+            }
+            $filtersParameters = $this->swaggerFilterDefinitions->get($swaggerMetadata);
+            if (count($filtersParameters) > 0) {
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: Useless count.

File: src/Swagger/Extractor/CollectionPostExtractor.php [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+namespace ApiPlatform\Core\Swagger\Extractor;
+
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
+use ApiPlatform\Core\Swagger\Model\SwaggerDocumentation;
+use ApiPlatform\Core\Swagger\Model\SwaggerMetadata;
+use ApiPlatform\Core\Swagger\Util\SwaggerDefinitions;
+use Symfony\Component\HttpFoundation\Request;
+
+class CollectionPostExtractor implements SwaggerExtractorInterface
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionPostExtractor.php]: final

File: src/Swagger/Extractor/CollectionPostExtractor.php [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+    public function extract(SwaggerMetadata $swaggerMetadata): SwaggerDocumentation
+    {
+        if (null !== $this->decoratedExtractor) {
+            $documentation = $this->decoratedExtractor->extract($swaggerMetadata);
+        } else {
+            $documentation = new SwaggerDocumentation();
+        }
+
+        if ($swaggerMetadata->isCollection() && in_array(Request::METHOD_POST, $swaggerMetadata->getMethods())) {
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionPostExtractor.php]: deps to HttpFoundation to remove

File: src/Swagger/Extractor/CollectionPostExtractor.php [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+            $resourceMetadata = $this->resourceMetadataFactory->create($swaggerMetadata->getResourceClass());
+
+            $this->extractSummary($resourceMetadata, $documentation);
+            $this->extractResponses($swaggerMetadata, $documentation, $resourceMetadata);
+            $this->extractParameters($swaggerMetadata, $documentation, $resourceMetadata);
+
+            if (empty($documentation->getConsumes())) {
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionPostExtractor.php]: useless empty call

File: src/Swagger/Extractor/CollectionPostExtractor.php [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+            $documentation->setSummary(sprintf('Creates a %s resource.', $resourceMetadata->getShortName()));
+        }
+    }
+
+    /**
+     * @param SwaggerMetadata $swaggerMetadata
+     * @param $documentation
+     * @param $resourceMetadata
+     */
+    protected function extractResponses(
```
Comment (dunglas) [inline src/Swagger/Extractor/CollectionPostExtractor.php]: We inline method prototypes (and we never use a protected method without a good use case)

**Comment by teohhanhui** [comment]
Interesting... Can you provide an overview of how it works? I'd like to see a value object for each "Swagger object", actually. And hopefully everything is immutable now.

**Comment by dunglas** [comment]
I'm fond of splitting the Swagger normalizer in smaller chunks of codes make it easier to extend! I'm fully with you on this topic, however I've some concern with the general design of the PR:

- I'm against introducing a new intermediate representation especially for Swagger (the `SwaggerMetadata` class), the main idea of API Platform 2 is to have a unique, generic representation of the API (the Metadata component) for all formats. Then, specific normalizer (Swagger, Hydra maybe RAML or API Blueprint afterward) must only use this generic set of metadata and serialize it in the supported format. (the component is already doing that, IMO it's a regression to change this behavior).
- I'm don't like using decorator for everything like here. IMO a cleaner alternative is to extract the logic to generate each type of operations in a specific class (to let the use decorate it if he has to) and inject those classes as dependencies of the main `SwaggerNormalizer` class. Something like:

```php
final class DocumentationNormalizer/* ... */
{
    public function __construct(GetOperationNormalizer $gon, PutOperationNormalizer $pon/* ... */)
```

What do you think?

**Comment by teohhanhui** [comment]
I'm also -1 on `SwaggerMetadata`. We should have Swagger-specific VOs instead, but that should never cross the layer boundary. In other words, the Swagger Object VOs should only be used in the (expanded / broken apart) Swagger DocumentationNormalizer.

**Comment by teohhanhui** [comment]
But hmm... Ideally these Swagger VOs should be reusable and hence best belong in another repository / package.

**Comment by dunglas** [comment]
Why not using associatives arrays instead of specific VOs? It would be easier to extend with custom properties, and "sub normalizers" can still implement the `NormalizerInterface` from the Symfony serializer.

**Comment by teohhanhui** [comment]
I'm okay with keeping to the current `ArrayObject` (remember we cannot use arrays, because there will be no way to tell between an empty object and an empty array, which is important for Swagger).

Not sure about the "sub-normalizers" implementing the Symfony Serializer's `NormalizerInterface` (what's the benefit?). I will be convinced when I see how it can be a good fit.

**Comment by dunglas** [comment]
I forgot about `\ArrayObject` but you're right. Even if we don't implement the `NormalizerInterface` itself (I've no strong opinion about that), keeping its signature allows to easily specialize the behavior using decoration.

**Comment by dunglas** [comment]
Double thinking about it, implementing an interface is mandatory to permit decoration.

**Comment by teohhanhui** [comment]
What I mean is, I'm not sure the same pattern as normalizer-sub normalizer applies here, because the specialized "extractors" would be working with the same `Documentation` object as input.

**Comment by piotrbrzezina** [comment]
Thanks for your answer and discussions 

I agree with you about that the decorator is not the best idea. I will try to extract the logic to generate each type of operations and inject those classes as dependencies of the main SwaggerNormalizer class.  
however in first approach I decide to do it as decorators was motivated motivated that the each normalizer (post/get/put..) have a lot of common element and I tried to extract this common elements to anther extracotr.

I'm a little confused when it comes to "SwaggerMetadata" I'm not sure what conclusions you have come? Should I use the Array Object or some other approach?

**Comment by dunglas** [comment]
Why should they? "sub normalizers" don't have to be aware of the other part of the docs.

**Comment by dunglas** [comment]
1. To reuse code, you can use traits or `@internal` helper class (with static methods) instead of decoration.
2. Yes, use an `ArrayObject`

Btw, can you try in a 1st time to do mostly a refactoring (splitting the current class), without changing too much the algorithms. It will ease the review process and prevent introducing regressions.

**Comment by piotrbrzezina** [comment]
Ok I will try to do small refactor first

File: src/Swagger/Extractor/CollectionGetExtractor.php (hunk 3 above)
Comment (mvrhov) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: Why do you find this useless.
I know that IDEs can resolve a lot more than back in the day but this at least to me sometimes scaning the properties is faster than looking at the long constructor.

Comment (dunglas) [inline src/Swagger/Extractor/CollectionGetExtractor.php]: IDE and static analysis tools infer the type from the constructor type hint. The Symfony CS (that we follow) tells to not add them (it eases the maintenance).

**Comment by piotrbrzezina** [comment]
@dunglas @teohhanhui 
What do you think about using tags services instead decoration ? 
I still prefer to  not add each post /get ... normalizer in constructor but pass one services that knows how to construct swagger documentation its easier to add new functions with out need to override anything

**Comment by teohhanhui** [comment]
I'd like to understand why service decoration is not good enough for your use case. It's only a matter of making parts of the SwaggerNormalizer more reusable, right?

**Comment by piotrbrzezina** [comment]
Hi
I finally finished making changes according to your suggestions.
Can you check it now ?

File: src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/SwaggerProcessorPass.php [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+     * @throws RuntimeException
+     */
+    public function process(ContainerBuilder $container)
+    {
+        // if doctrine not loaded
+        if (!$container->hasDefinition('api_platform.swagger.processor.swagger_extractor_processor')) {
+            return;
+        }
+        $extractors = $this->findAndSortTaggedServices('api_platform.swagger_extractor', $container);
+        $container->getDefinition('api_platform.swagger.processor.swagger_extractor_processor')->addArgument($extractors);
```
Comment (Simperfit) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/SwaggerProcessorPass.php]: maybe this `$extractors` can be directly inlined ?

Comment (piotrbrzezina) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/SwaggerProcessorPass.php]: Done

**Comment by piotrbrzezina** [comment]
@teohhanhui I think decoration service is good enough to make "SwaggerNormalizer" more reusable. But in my opinion it easier to add new swagger extractor extension by making new services with tags „api_platform.swagger_extractor”. I'm already implemented this approach what do you think about it ?

@dunglas @Simperfit 
if you have a few minutes,please review this PR

**Pull request closed without merging**